cl/api/pagination.py

```diff
@@ -16,19 +21,107 @@
 
 
-class ShallowOnlyPageNumberPagination(PageNumberPagination):
-    """A paginator that blocks deep pagination
-
-    Thank you MuckRock for this contribution.
+class VersionBasedPagination(PageNumberPagination):
+    """The base paginator for handling V3 and V4 DB endpoints.
+    This supports CursorPagination for V4 endpoints when sorting by "id" or
+    "date_created". It uses PageNumberPagination for V3 endpoints and for V4
+    endpoints when sorting by keys that don't support CursorPagination.
     """
 
     max_pagination_depth = 100
+    version = "v3"
+    cursor_query_param = "cursor"
+    invalid_cursor_message = "Invalid cursor"
+    compatible_sorting = {
+        "id": "int",
+        "-id": "int",
+        "date_created": "date",
+        "-date_created": "date",
+        "date_modified": "date",
+        "-date_modified": "date",
+    }
+    ordering = ""
+    other_cursor_ordering_keys = []
+
+    def __init__(self):
+        super().__init__()
+        self.cursor_paginator = CursorPagination()
+        self.cursor_paginator.page_size = self.page_size
+
+    def do_v4_cursor_pagination(self):
+        """Determine if v4 cursor pagination should be applied.
+
+        :return: A two tuple containing:
+        - A boolean indicating if cursor pagination should be applied.
+        - The requested ordering key if applicable.
+        """
+
+        all_cursor_ordering_keys = []
+        requested_ordering = self.request.query_params.get(
+            "order_by", self.ordering
+        )
+        all_cursor_ordering_keys.extend(self.other_cursor_ordering_keys)
+        all_cursor_ordering_keys.append(self.ordering)
+        return (
+            all(
+                [
+                    self.version == "v4",
+                    requested_ordering,
+                    requested_ordering in all_cursor_ordering_keys,
+                ]
+            ),
+            requested_ordering,
+        )
 
     def paginate_queryset(self, queryset, request, view=None):
         """
         Paginate a queryset if required, either returning a
         page object, or `None` if pagination is not configured for this view.
         """
+
+        if hasattr(view, "ordering"):
+            self.ordering = view.ordering
+        if hasattr(view, "other_cursor_ordering_keys"):
+            self.other_cursor_ordering_keys = view.other_cursor_ordering_keys
+
+        self.version = request.version
+        self.request = request
+        do_cursor_pagination, requested_ordering = (
+            self.do_v4_cursor_pagination()
+        )
+        if do_cursor_pagination:
+            # Handle the queryset using CursorPagination
+            return handle_database_cursor_pagination(
+                self, request, requested_ordering, queryset, view
+            )
+
+        # Handle the queryset using PageNumberPagination
+        return self.handle_shallow_only_page_number_pagination(
+            request, queryset
+        )
```

Comment:
It's a little weird that shallow pagination is a method and cursor pagination is a function. Fixable?

Reply:
Yeah fixed. The problem was related to mocking `handle_database_cursor_pagination` in the tests. Using mock as a class method with `side_effect` caused a recursion loop. I found a way to keep the tests working while moving `handle_database_cursor_pagination` to a class method. The solution involved using `new` instead of `side_effect` in the mock and applying additional properties to the mock method to maintain the same assertions as before.